oidcccl,provisioning: add user provisioning for OIDC authentication

shriramters opened the pull request:
Previously, the OIDC authentication flow in the DB Console only supported logging in with existing user accounts. There was no built-in mechanism to automatically create a new database user when a user authenticated via an OIDC provider for the first time.

This was inadequate for organizations that wanted to streamline user onboarding and reduce manual administrative overhead. Every new user needing access to the database had to have a corresponding user account created manually by a database administrator, which is both time-consuming and prone to error.

To address this, this patch introduces automatic user provisioning for the OIDC authentication flow. When a new user successfully authenticates via an OIDC provider, a corresponding CockroachDB user is now automatically created if one does not already exist. This functionality is controlled by a new cluster setting, `security.provisioning.oidc.enabled`, which is disabled by default to maintain backward compatibility and ensure administrators can opt-in to this behavior.

Fixes: #126680
Epic: CRDB-48764

Release note (enterprise change): A new cluster setting, `security.provisioning.oidc.enabled`, has been added to allow for the automatic provisioning of users when they log in for the first time via OIDC. When enabled, a new user will be created in CockroachDB upon their first successful OIDC authentication. This feature is disabled by default.

shriramters committed:
oidcccl,provisioning: add user provisioning for OIDC authentication

Previously, the OIDC authentication flow in the DB Console only
supported logging in with existing user accounts. There was no
built-in mechanism to automatically create a new database user when a
user authenticated via an OIDC provider for the first time.

This was inadequate because administrators would need to manually
create a database user for every first‑time OIDC login, adding
friction and overhead to onboarding.

To address this, this patch introduces automatic user provisioning for
the OIDC authentication flow. When a new user successfully
authenticates via an OIDC provider, a corresponding CockroachDB user
is now automatically created if one does not already exist. This
functionality is controlled by a new cluster setting,
`security.provisioning.oidc.enabled`, which is disabled by default to
maintain backward compatibility and ensure administrators can opt-in
to this behavior.

Fixes: #126680
Epic: CRDB-48764

Release note (enterprise change): A new cluster setting,
`security.provisioning.oidc.enabled`, has been added to allow for the
automatic provisioning of users when they log in for the first time
via OIDC. When enabled, a new user will be created in CockroachDB upon
their first successful OIDC authentication. This feature is disabled
by default.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/150687)
<!-- Reviewable:end -->

shriramters commented on pkg/ccl/oidcccl/authentication_oidc_test.go:
Good catch, I copied and pasted the previous test and changed this to false

shriramters commented on pkg/ccl/oidcccl/authentication_oidc_test.go:
```diff
@@ -91,7 +92,7 @@ func (m mockOidcManager) ExchangeVerifyGetTokenInfo(
 	ctx context.Context, code, idTokenKey string, _ bool,
 ) (map[string]json.RawMessage, map[string]json.RawMessage, string, string, error) {
 	claims := map[string]json.RawMessage{
-		"email": json.RawMessage(`"test@example.com"`),
+		"email": json.RawMessage(fmt.Sprintf(`"%s"`, m.claimEmail)),
```
Agreed

souravcrl reviewed (commented):
> Reviewed 6 of 7 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150687)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @shriramters)

___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 627 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-9bHAC0zrQenFt86e:-OW-9bHAC0zrQenFt86f:b493r9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc.go#L627)):*
errors.Wrapf will mark the IDP string as unsafe for logging:, refer: https://github.com/cockroachdb/errors/blob/2008f7c3ac42391b26a4bbb1cf0bbf7057605610/errutil_api.go#L116-L132
To avoid this and since we don't need the original error for posterity, we could simple create a new formatted error while logging on the line below:

> log.Errorf(ctx, "OIDC provisioning: error parsing provisioning source IDP %s: %v", idString, err)
___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 636 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-FwV824Oh04yK4H9b:-OW-FwV824Oh04yK4H9c:bojkasf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc.go#L636)):*
errors.Wrapf will mark the IDP string as unsafe for logging:, refer: https://github.com/cockroachdb/errors/blob/2008f7c3ac42391b26a4bbb1cf0bbf7057605610/errutil_api.go#L116-L132
To avoid this and since we don't need the original error for posterity, we could simple create a new formatted error while logging on the line below:

> log.Errorf(ctx, "OIDC provisioning: error provisioning user %s: %v", sqlUsername, err)
___
*[`-- commits` line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OVxELUt1u2DVN7W_UaR:-OVxELUt1u2DVN7W_UaS:b45fl7f):*
This comment block in this paragraph does not seem relevant in the git commit message. The commit message is intended to communicate context about the change to fellow developers working on the codebase to make it easier to maintain. This seems part of the problem statement itself to communicate the gap in the product and should ideally go in the GH issue. We may need to prune it or skip it altogether.
___
*[`pkg/security/provisioning/settings.go` line 19 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-Gel-Cgii9wzV_6RB:-OW-Gel-Cgii9wzV_6RC:b8hvu2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/security/provisioning/settings.go#L19)):*
> ```Go
> const (
> 	supportedAuthMethodLDAP             = "ldap"
> 	supportedAuthMethodOIDC             = "oidc"
> ```

To note: OIDC is not a valid auth method for db logins and is not registered with `AuthMethods` interface. Could we add a comment to the effect that although `oidc` is not a valid method for database auth, to maintain parity with other db auth methods supporting provisioning like ldap/jwt_token, we will be a defining a oidc auth method which could be accessed through the `PROVISIONSRC` role option set for the provisioned user. see: https://github.com/cockroachdb/cockroach/blob/4f2cf60586dea80c4246d611cb700c62af35110c/pkg/sql/pgwire/auth_methods.go#L89-L99
___
*[`pkg/ccl/oidcccl/authentication_oidc_test.go` line 927 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-M6XrBggx2ULVWb5E:-OW-M6XrBggx2ULVWb5F:b-uwyps2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc_test.go#L927)):*
> ```Go
> 		require.False(t, rows.Next(), "user should not have been created")
> 		require.NoError(t, rows.Close())
> 	})
> ```

Just to test the initial hypothesis can we test for provisioning disabled, existing user?

Additionally can we have a test 
1. where the returned token fails the verification in ExchangeVerifyGetTokenInfo. 
2. where issuer field does not match the set oidcAuthentication.conf.providerURL(assuming provider is also the issuer of the token)
3. where issuer field does not match the set oidcAuthentication.conf.providerURL(assuming provider is not the the issuer of the token). How do we validate the source in this case?
___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-Iug8AFCvfMLLL_Fn:-OW-Iug8AFCvfMLLL_Fo:b-s3yjqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc.go#L58)):*
> ```Go
> 	beginJWTCounterName      = counterPrefix + "begin_jwt"
> 	loginSuccessCounterName  = counterPrefix + "login_success"
> 	enableCounterName        = counterPrefix + "enable"
> ```

We will need enablement and success counters for oidc provisioning as well
___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 612 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-D5v-2wwxAsGKlsP2:-OW-D5v-2wwxAsGKlsP3:bnx1rmf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc.go#L612)):*
> ```Go
> 		}
> 
> 		provConfig := provisioning.ClusterProvisioningConfig(st)
> ```

To improve code maintainability, we can move the provisioning logic to a separate utility including the validation of `oidcAuthenticationConf.provisioningEnabled`, check my comment below.
___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 613 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-DVcB7ciIN3ngP4VN:-OW-DVcB7ciIN3ngP4VO:b-64hn4p) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc.go#L613)):*
> ```Go
> 
> 		provConfig := provisioning.ClusterProvisioningConfig(st)
> 		if provConfig.Enabled("oidc") {
> ```

We need move this to `oidcAuthenticationConf` https://github.com/cockroachdb/cockroach/blob/d1d049a9951744070342bc09275cdb88ce36a658/pkg/ccl/oidcccl/authentication_oidc.go#L136-L155 

and set an reload event for change like we do for oidcEnabled cluster setting: https://github.com/cockroachdb/cockroach/blob/d1d049a9951744070342bc09275cdb88ce36a658/pkg/ccl/oidcccl/authentication_oidc.go#L784-L786?
___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 623 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-IYJc16S9hUVPSWAv:-OW-IYJc16S9hUVPSWAw:b-c9vr27) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc.go#L623)):*
> ```Go
> 
> 			// Create the provisioning source identifier string, e.g., "oidc:https://accounts.example.com".
> 			idpString := "oidc:" + oidcAuthentication.conf.providerURL
> ```

should we use `id_token` as the proxy auth method here? We can reuse the constant previously defined https://github.com/cockroachdb/cockroach/blob/d1d049a9951744070342bc09275cdb88ce36a658/pkg/ccl/oidcccl/authentication_oidc.go#L41 or if we want to use `oidc` as the provisioning method, we can define a new constant for consistency.
___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 624 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-MzEf2yx9PGOvI8J9:-OW-MzEgCVqYBevU7ENW:bd5dwr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc.go#L624)):*
> ```Go
> 			// Create the provisioning source identifier string, e.g., "oidc:https://accounts.example.com".
> 			idpString := "oidc:" + oidcAuthentication.conf.providerURL
> 			provisioningSource, err := provisioning.ParseProvisioningSource(idpString)
> ```

Should we validate the oidcAuthentication.conf.providerURL string first with issuer field in the token and do we need a validation for JWKS?
___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 640 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-GExw7zoNxPW0CIOw:-OW-GExw7zoNxPW0CIOx:b-udedaj) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc.go#L640)):*
> ```Go
> 				return
> 			}
> 			log.Infof(ctx, "OIDC: successfully provisioned user %s", sqlUsername)
> ```

This seems inconsistent with the logging in this package. Would be ideal to provide a a begin provisioning INFO log as well.


<!-- Sent from Reviewable.io -->

shriramters reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150687)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @souravcrl)

___
*[`-- commits` line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OVxELUt1u2DVN7W_UaR:-OW0wQwRA-bNAsEhrGEr:bm0s2w0):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

This comment block in this paragraph does not seem relevant in the git commit message. The commit message is intended to communicate context about the change to fellow developers working on the codebase to make it easier to maintain. This seems part of the problem statement itself to communicate the gap in the product and should ideally go in the GH issue. We may need to prune it or skip it altogether.
</blockquote></details>

Agreed. I was actually just following the git commit template which needed a paragraph starting with "This was inadequate.." I didn't know what exactly is expected there.

I have changed it now. Please tell me if this looks okay.
___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-Iug8AFCvfMLLL_Fn:-OW0urbv761gl3bpRqLy:bur8ib6) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc.go#L58)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

We will need enablement and success counters for oidc provisioning as well
</blockquote></details>

I believe this should be added in https://reviewable.io/reviews/cockroachdb/cockroach/150687#file-OVrRK_Ju-bVSfCK-MVg

Done. Added it there because the provisioning telemetry for all the other auth methods is added there
___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 612 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-D5v-2wwxAsGKlsP2:-OW0tbrC6taNTa5Qr8Jo:b-qg5m5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc.go#L612)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

To improve code maintainability, we can move the provisioning logic to a separate utility including the validation of `oidcAuthenticationConf.provisioningEnabled`, check my comment below.
</blockquote></details>

Done, moved it to `provisionUserIfEnabled()`
___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 613 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-DVcB7ciIN3ngP4VN:-OW0mqLl5qfRNG8O5Cw8:b-mwv2hy) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc.go#L613)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

We need move this to `oidcAuthenticationConf` https://github.com/cockroachdb/cockroach/blob/d1d049a9951744070342bc09275cdb88ce36a658/pkg/ccl/oidcccl/authentication_oidc.go#L136-L155 

and set an reload event for change like we do for oidcEnabled cluster setting: https://github.com/cockroachdb/cockroach/blob/d1d049a9951744070342bc09275cdb88ce36a658/pkg/ccl/oidcccl/authentication_oidc.go#L784-L786?
</blockquote></details>

Done. Moved to `oidcAuthentication.conf.provisioningEnabled`
___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 623 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-IYJc16S9hUVPSWAv:-OW0eNXL57WXpFb98FK7:b-7n70s1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc.go#L623)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

should we use `id_token` as the proxy auth method here? We can reuse the constant previously defined https://github.com/cockroachdb/cockroach/blob/d1d049a9951744070342bc09275cdb88ce36a658/pkg/ccl/oidcccl/authentication_oidc.go#L41 or if we want to use `oidc` as the provisioning method, we can define a new constant for consistency.
</blockquote></details>

added a constant `oidcProvisioningKey`
___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 624 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-MzEf2yx9PGOvI8J9:-OW0fA8J0rhHQLksb73k:b-ow8o2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc.go#L624)):*
I believe this is not necessary

> do we need a validation for JWKS?

`go-oidc` validates the JWKS, During discovery the `oidc.Provider` downloads `/.well‑known/openid‑configuration` from where JWKS is obtained

> Should we validate the oidcAuthentication.conf.providerURL string first with issuer field in the token

This runs after `ExchangeVerifyGetTokenInfo`, which has called the `go-oidc` verifier. At this point all these are verified:
- issuer
- audience and expiry
- signature (and in turn JWKS)

So we can safely trust `idpString` (which derived from `iss`) to map the login to a provisioning source.
___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 627 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-9bHAC0zrQenFt86e:-OW0hqwB0jzoMUOM-A9H:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc.go#L627)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

errors.Wrapf will mark the IDP string as unsafe for logging:, refer: https://github.com/cockroachdb/errors/blob/2008f7c3ac42391b26a4bbb1cf0bbf7057605610/errutil_api.go#L116-L132
To avoid this and since we don't need the original error for posterity, we could simple create a new formatted error while logging on the line below:

> log.Errorf(ctx, "OIDC provisioning: error parsing provisioning source IDP %s: %v", idString, err)
</blockquote></details>

Done
___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 636 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-FwV824Oh04yK4H9b:-OW0i0E174OMONuwbS5h:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc.go#L636)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

errors.Wrapf will mark the IDP string as unsafe for logging:, refer: https://github.com/cockroachdb/errors/blob/2008f7c3ac42391b26a4bbb1cf0bbf7057605610/errutil_api.go#L116-L132
To avoid this and since we don't need the original error for posterity, we could simple create a new formatted error while logging on the line below:

> log.Errorf(ctx, "OIDC provisioning: error provisioning user %s: %v", sqlUsername, err)
</blockquote></details>

Done
___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 640 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-GExw7zoNxPW0CIOw:-OW0ZIM73xdTa1ZgjFaV:b-4w4x0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc.go#L640)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

This seems inconsistent with the logging in this package. Would be ideal to provide a a begin provisioning INFO log as well.
</blockquote></details>

Added a info log at the beginning 



_Code snippet:_
```Go
log.Infof(ctx, "OIDC: attempting user provisioning for %s", username)
```
___
*[`pkg/security/provisioning/settings.go` line 19 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-Gel-Cgii9wzV_6RB:-OW0b52o8T1QWfPlSw-2:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/security/provisioning/settings.go#L19)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

To note: OIDC is not a valid auth method for db logins and is not registered with `AuthMethods` interface. Could we add a comment to the effect that although `oidc` is not a valid method for database auth, to maintain parity with other db auth methods supporting provisioning like ldap/jwt_token, we will be a defining a oidc auth method which could be accessed through the `PROVISIONSRC` role option set for the provisioned user. see: https://github.com/cockroachdb/cockroach/blob/4f2cf60586dea80c4246d611cb700c62af35110c/pkg/sql/pgwire/auth_methods.go#L89-L99
</blockquote></details>

Done
___
*[`pkg/ccl/oidcccl/authentication_oidc_test.go` line 927 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-M6XrBggx2ULVWb5E:-OW0Oxby5RLdJpPbi5V3:bymu2m6) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc_test.go#L927)):*
> where issuer field does not match the set oidcAuthentication.conf.providerURL(assuming provider is not the the issuer of the token). How do we validate the source in this case?

Generally this is the federated oidc case but cockroach docs explicitly prohibit this.

From the docs:
>  the issuer URL must match the value of `server.oidc_authentication.provider_url`.

source: https://www.cockroachlabs.com/docs/v25.3/sso-sql#:~:text=the%20issuer%20URL%20must%20match%20the%20value%20of%20server.oidc_authentication.provider_url.

I added tests enforcing these. These tests add ~250 lines as they need specific overriding of the mock functions.


<!-- Sent from Reviewable.io -->

souravcrl reviewed (commented):
> Reviewed 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150687)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @shriramters)

___
*[`-- commits` line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OVxELUt1u2DVN7W_UaR:-OW56aOE653lGOEkDiwT:b9qa6gy):*
<details><summary><i>Previously, shriramters (Shriram Ravindranathan) wrote…</i></summary><blockquote>

Agreed. I was actually just following the git commit template which needed a paragraph starting with "This was inadequate.." I didn't know what exactly is expected there.

I have changed it now. Please tell me if this looks okay.
</blockquote></details>

This seems better now. Thank you for incorporating the change.
___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 624 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-MzEf2yx9PGOvI8J9:-OW6EE8V2N72V1Babx32:b-qqevvb) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc.go#L624)):*
> During discovery the `oidc.Provider` downloads `/.well‑known/openid‑configuration` from where JWKS is obtained

This assumption is only true for providers which expose the openid-configuration and can also be disabled for security reasons.

> This runs after `ExchangeVerifyGetTokenInfo`, which has called the `go-oidc` verifier. At this point all these are verified:
>
> -   issuer

This is also a known assumption and while we don't validate this in the code we also don't explicitly mention this assumption incase it is violated, see: https://github.com/coreos/go-oidc/blob/a7c457eacb849c163a496b29274242474a8f44ab/oidc/verify.go#L260-L269

Can we add a comment for both these assumptions for reference since we have encountered issues with the setup in JWT authentication and had to provide additional configurations to handle the issue.
___
*[`pkg/security/provisioning/settings.go` line 19 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-Gel-Cgii9wzV_6RB:-OW6GoisEaSd3y41Ijhs:bh2rgqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/security/provisioning/settings.go#L19)):*
<details><summary><i>Previously, shriramters (Shriram Ravindranathan) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Can you please also add a `TODO(souravcrl): ensure proper setting of the role option for PROVISIONSRC incase both JWT and OIDC provisioning are enabled. While users can be provisioned post authentication from both auth methods, since the underlying user identity is the JWT token, we can end up with divergent user cohorts due to auth method set for the option value.`
___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 501 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW57DZv2POY9L282vRG:-OW57DZv2POY9L282vRH:b-7eczj2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac84f4d9da1e6bdee521642676197cd88d493495/pkg/ccl/oidcccl/authentication_oidc.go#L501)):*
> ```Go
> // user linked to the OIDC identity provider. Any error encountered is logged
> // and sent back to the client as a 500 response.
> func provisionUserIfEnabled(
> ```

in keeping with the naming standards in the package, lets rename it to `maybeProvisionUserLocked` as it expects `oidcAuthenticationConf` to be in locked state and conditionally provisions the user subject to enablement of the cluster and other dependent criteria being satisfied.
___
*[`pkg/ccl/oidcccl/authentication_oidc_test.go` line 1072 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW6NF175N5MFbyF8gLd:-OW6NF175N5MFbyF8gLe:b4njb6a) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac84f4d9da1e6bdee521642676197cd88d493495/pkg/ccl/oidcccl/authentication_oidc_test.go#L1072)):*
> ```Go
> // but forces issuer verification to fail.
> type issuerMismatchOidcManager struct {
> 	mockOidcManager
> ```

The boilerplate code seems unnecessary, can you update the mockOIDCManager with additional config values which are validated in the type definitions itself, see: https://github.com/cockroachdb/cockroach/blob/20e61e684f7f1b5c4337ccd2bbe831b8ae9eb712/pkg/ccl/oidcccl/authentication_oidc_test.go#L71-L74

```
type mockOidcManager struct {
	oauth2Config *oauth2.Config
	claimEmail   string
        issuerMismatch bool
}
```

and then validate if the config is set in the `Verify`/`ExchangeVerifyGetTokenInfo`.
___
*[`pkg/ccl/oidcccl/authentication_oidc_test.go` line 1122 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW6OyCoEutMCfiMb9gL:-OW6OyCpCTSmqSSmCvUq:bypvf01) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac84f4d9da1e6bdee521642676197cd88d493495/pkg/ccl/oidcccl/authentication_oidc_test.go#L1122)):*
> ```Go
> 	ctx context.Context, code, idTokenKey string, _ bool,
> ) (map[string]json.RawMessage, map[string]json.RawMessage, string, string, error) {
> 	// go-oidc will throw an error unless SkipIssuerCheck is configured
> ```

This seems inconsistent with your comment https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-MzEf2yx9PGOvI8J9 , and although it's true, it doesn't seem to be written by you. Please only submit code after you have validated it yourself.
___
*[`pkg/ccl/oidcccl/authentication_oidc_test.go` line 1136 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW6PyQs3tojX-L0errn:-OW6PyQs3tojX-L0erro:bsmx9wv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac84f4d9da1e6bdee521642676197cd88d493495/pkg/ccl/oidcccl/authentication_oidc_test.go#L1136)):*
> ```Go
> func TestOIDCIssuerValidation(t *testing.T) {
> 	// Sub‑test 1: provider is issuer; mismatch must fail.
> 	t.Run("issuer mismatch fails when provider is issuer", func(t *testing.T) {
> ```

This should succeed right since provider is indeed issuer. Why are we failing this? Did you validate the test case before submitting?
___
*[`pkg/ccl/oidcccl/authentication_oidc_test.go` line 1265 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW6QkuR4PlB-0wo2Y2g:-OW6QkuR4PlB-0wo2Y2h:b-bjxmc7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac84f4d9da1e6bdee521642676197cd88d493495/pkg/ccl/oidcccl/authentication_oidc_test.go#L1265)):*
> ```Go
> 			cbReq.AddCookie(c)
> 		}
> 		cbResp, err := httpClient.Do(cbReq)
> ```

Can we validate the response body to be the error we are sending?


<!-- Sent from Reviewable.io -->

shriramters commented:
*[`pkg/ccl/oidcccl/authentication_oidc_test.go` line 1122 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW6OyCoEutMCfiMb9gL:-OW93l4b5vfKnXb072dD:b-q9bv6w) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac84f4d9da1e6bdee521642676197cd88d493495/pkg/ccl/oidcccl/authentication_oidc_test.go#L1122)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

This seems inconsistent with your comment https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-MzEf2yx9PGOvI8J9 , and although it's true, it doesn't seem to be written by you. Please only submit code after you have validated it yourself.
</blockquote></details>

I'm sorry about the confusion but this comment was actually written by me.

I added this only in the latest revision after I discovered this line from the `go-oidc` source https://github.com/coreos/go-oidc/blob/a7c457eacb849c163a496b29274242474a8f44ab/oidc/verify.go#L260

I added it to indicate why `ExchangeVerifyGetTokenInfo` is mocked to fail in this case.

My earlier assumption was that it should fail because cockroach docs said so (As per my other comment). I will be more thorough next time.


<!-- Sent from Reviewable.io -->

shriramters commented:
*[`pkg/ccl/oidcccl/authentication_oidc_test.go` line 1136 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW6PyQs3tojX-L0errn:-OW95yDc69sdzXdMB405:b17ficy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac84f4d9da1e6bdee521642676197cd88d493495/pkg/ccl/oidcccl/authentication_oidc_test.go#L1136)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

This should succeed right since provider is indeed issuer. Why are we failing this? Did you validate the test case before submitting?
</blockquote></details>

Sorry about the confusing wording. Now that I think of it, the sentence doesn't make sense at all at first glance

What I meant was:

"When provider is issuer" =  the cluster setting provider_url matches the issuer retrieved from `/.well‑known/openid‑configuration`

"issuer mismatch fails" = ID Token has different `iss` from those two values. This "issuer mismatch" will "fail" the test.


<!-- Sent from Reviewable.io -->

shriramters reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150687)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @souravcrl)

___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 624 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-MzEf2yx9PGOvI8J9:-OWAIJpc2aeFrbx2KQj5:b-yso0md) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/ccl/oidcccl/authentication_oidc.go#L624)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

> During discovery the `oidc.Provider` downloads `/.well‑known/openid‑configuration` from where JWKS is obtained

This assumption is only true for providers which expose the openid-configuration and can also be disabled for security reasons.

> This runs after `ExchangeVerifyGetTokenInfo`, which has called the `go-oidc` verifier. At this point all these are verified:
>
> -   issuer

This is also a known assumption and while we don't validate this in the code we also don't explicitly mention this assumption incase it is violated, see: https://github.com/coreos/go-oidc/blob/a7c457eacb849c163a496b29274242474a8f44ab/oidc/verify.go#L260-L269

Can we add a comment for both these assumptions for reference since we have encountered issues with the setup in JWT authentication and had to provide additional configurations to handle the issue.
</blockquote></details>

I've added these assumptions to the doc comment above `maybeProvisionUserLocked`
___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 501 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW57DZv2POY9L282vRG:-OWAFASE8sTNjwM1SsI3:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac84f4d9da1e6bdee521642676197cd88d493495/pkg/ccl/oidcccl/authentication_oidc.go#L501)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

in keeping with the naming standards in the package, lets rename it to `maybeProvisionUserLocked` as it expects `oidcAuthenticationConf` to be in locked state and conditionally provisions the user subject to enablement of the cluster and other dependent criteria being satisfied.
</blockquote></details>

Done
___
*[`pkg/security/provisioning/settings.go` line 19 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW-Gel-Cgii9wzV_6RB:-OWAIuIP7t5VPSyc1Ge1:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/6556a86483b7b684ce7b4d2e7199f940a15c2f10/pkg/security/provisioning/settings.go#L19)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

Can you please also add a `TODO(souravcrl): ensure proper setting of the role option for PROVISIONSRC incase both JWT and OIDC provisioning are enabled. While users can be provisioned post authentication from both auth methods, since the underlying user identity is the JWT token, we can end up with divergent user cohorts due to auth method set for the option value.`
</blockquote></details>

Done
___
*[`pkg/ccl/oidcccl/authentication_oidc_test.go` line 1072 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW6NF175N5MFbyF8gLd:-OWAOYfN9JGDdmi06suY:b-a3sl72) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac84f4d9da1e6bdee521642676197cd88d493495/pkg/ccl/oidcccl/authentication_oidc_test.go#L1072)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

The boilerplate code seems unnecessary, can you update the mockOIDCManager with additional config values which are validated in the type definitions itself, see: https://github.com/cockroachdb/cockroach/blob/20e61e684f7f1b5c4337ccd2bbe831b8ae9eb712/pkg/ccl/oidcccl/authentication_oidc_test.go#L71-L74

```
type mockOidcManager struct {
	oauth2Config *oauth2.Config
	claimEmail   string
        issuerMismatch bool
}
```

and then validate if the config is set in the `Verify`/`ExchangeVerifyGetTokenInfo`.
</blockquote></details>

Thanks so much, this reduced like 70 lines of code. So much cleaner!
___
*[`pkg/ccl/oidcccl/authentication_oidc_test.go` line 1265 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-OW6QkuR4PlB-0wo2Y2g:-OWAOJVA-7pD47hrs8dk:b1r5gdh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac84f4d9da1e6bdee521642676197cd88d493495/pkg/ccl/oidcccl/authentication_oidc_test.go#L1265)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

Can we validate the response body to be the error we are sending?
</blockquote></details>

Done. We are sending back `genericCallbackHTTPError` = "OIDC: unable to complete authentication". I have added a check to verify this is in the body.

(added to previous test also)


<!-- Sent from Reviewable.io -->

shriramters reviewed (commented):
@souravcrl 

After some thought, I have realized that the second test is virtually identical to the first one. The only difference is the string value used for the `OIDCProviderURL` cluster setting. However, because the tests use a mock that completely bypasses the `go-oidc` library's provider discovery so we can't test it. 

This was due to my initial misconception that we were specifically ensuring provider_url = iss claim in the code as per the cockroach docs.

I propose that we remove the second test.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150687#-:-OWAUpWY6GEiqXIcevqU:bfmrd83)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @souravcrl)



<!-- Sent from Reviewable.io -->
